rmf_websocket/src/rmf_websocket/client/ClientWebSocketEndpoint.cpp

```diff
@@ -2,66 +2,92 @@
 #include <chrono>
 #include <mutex>
 #include <sstream>
+#include <thread>
 
 using namespace std::chrono_literals;
 using namespace rmf_websocket;
-
+using namespace boost;
 
 //=============================================================================
 ConnectionMetadata::ConnectionMetadata(
-  websocketpp::connection_hdl hdl, std::string uri)
+  websocketpp::connection_hdl hdl, std::string uri, ConnectionCallback cb,
+  ReconnectionCallback rcb)
 : _hdl(hdl)
-  , _status("Connecting")
+  , _status(ConnectionStatus::CONNECTING)
   , _uri(uri)
   , _server("N/A")
+  , _connection_cb(cb)
+  , _reconnection_cb(std::move(rcb))
+{
+}
+
+//=============================================================================
+void ConnectionMetadata::reset()
 {
+  _status = ConnectionStatus::CONNECTING;
 }
 
 //=============================================================================
 void ConnectionMetadata::on_open(WsClient* c, websocketpp::connection_hdl hdl)
 {
-  {
-    std::lock_guard<std::mutex> lock(_status_mtx);
-    _status = "Open";
-    WsClient::connection_ptr con = c->get_con_from_hdl(hdl);
-    _server = con->get_response_header("Server");
-  }
-  _cv.notify_all();
+  _status = ConnectionStatus::OPEN;
+  WsClient::connection_ptr con = c->get_con_from_hdl(hdl);
+  _server = con->get_response_header("Server");
+  _connection_cb();
 }
 
 //=============================================================================
 void ConnectionMetadata::on_fail(WsClient* c, websocketpp::connection_hdl hdl)
 {
-  {
-    std::lock_guard<std::mutex> lock(_status_mtx);
-    _status = "Failed";
+  _status = ConnectionStatus::FAILED;
 
-    WsClient::connection_ptr con = c->get_con_from_hdl(hdl);
-    _server = con->get_response_header("Server");
-    _error_reason = con->get_ec().message();
-  }
-  _cv.notify_all();
+  WsClient::connection_ptr con = c->get_con_from_hdl(hdl);
+  _server = con->get_response_header("Server");
+  _error_reason = con->get_ec().message();
+  _reconnection_cb();
```

Comment:
We should `dispatch` the cb, this queues it at the back of the event loop, allowing potential cleanup to be done properly (even if websocketpp doesn't need it, it's still a good practice in general).